Defedit 1072 faster tests

RagnarSvenssonKing opened the pull request:
# Tech details

The tests are sped up by providing a function which memoizes project loading. In order for this to work, the-system needs to be copyable. Most of the changes are related to removing mutables etc and simplifying internal.system and project-building (lots of mutables) so that the-system is not changed by tests which would affect subsequent tests. More details in the comments in the diffs.

# Dev facing changes

- lein test takes ~ 3 mins rather than ~ 7 mins to run (on my machine)
- editor2 on CI went from 11 mins, 58 secs to 8 mins, 39 secs
- some tests are basically instantaneous to run from REPL, e.g. sprite-test (not the first time though)
-  systems are simpler to understand, especiall internal.system and project building

# User facing changes

- Building a project requires much less memory, as we write binary content to disk immediately as it is produced, rather than storing the entire project in memory before writing it to disk

RagnarSvenssonKing committed:
DEFEDIT-1072 [WIP] Project loading is cached through memoized

* Greatly speeds up tests
* Some bug remains with undo-test/redo-undone-deletion-still-deletes
* Also build-test does not run due to build-caches

RagnarSvenssonKing committed:
Cleaned up system ns API

RagnarSvenssonKing committed:
More API cleanup

RagnarSvenssonKing committed:
Further API cleanup

RagnarSvenssonKing committed:
Moved history out from graph and into 'system' map

RagnarSvenssonKing committed:
Moved cache mgmt partially from dynamo.graph to internal.system

RagnarSvenssonKing committed:
Moved cache mgmt from dynamo.graph to internal.system

RagnarSvenssonKing committed:
Even more API cleanup between dynamo.graph and internal.system

RagnarSvenssonKing committed:
Moved cache mgmt from internal.node to internal.system, partially

RagnarSvenssonKing committed:
Moved more cache mgmt from internal.node to internal.system

RagnarSvenssonKing committed:
Caches are now values

* System ref's it to deal with synchronization
* Others can only get the value
* Temp. disabled cache-syncing after long-running ops (save/build)
* Patched tests

RagnarSvenssonKing committed:
Fixed one more cache deref when building project

RagnarSvenssonKing committed:
g/user-data to store mutable user-data for nodes

RagnarSvenssonKing committed:
All build caches removed from defold-project, replaced by one file-based in pipeline ns

* Also handles etags
* New cache based on g/user-data rather than atoms

RagnarSvenssonKing committed:
Removed atom from CameraController node

* Replaced by g/user-data

RagnarSvenssonKing committed:
Merge branch 'DEFEDIT-1072-faster-tests' into DEFEDIT-1072-attempt-2

RagnarSvenssonKing committed:
build-test using with-loaded-project

RagnarSvenssonKing committed:
defold-project-search-test using test-util/with-loaded-project

RagnarSvenssonKing committed:
Removed expensive assert-less perf test

RagnarSvenssonKing committed:
Made gui-test ns reloadable

RagnarSvenssonKing committed:
Parts of gui-test extracted into perf-test, which was also made to run faster

* Larger perf threshold since less time spent on JIT optimizations

RagnarSvenssonKing committed:
Removed benchmark-test

* No assertions and not used for anything practical
* Also took a lot of time to run

RagnarSvenssonKing committed:
Merge remote-tracking branch 'origin/dev' into DEFEDIT-1072-attempt-2

matgis-king reviewed (changes requested):
Reviewed. Epic work. I am however concerned about the  change to caching behavior for `g/node-value`.

erikangelin-king committed:
Merge branch 'dev' into DEFEDIT-1072-faster-tests

erikangelin-king committed:
DEFEDIT-1072 Review fixes

- only update the system cache if graph unchanged
- artefact -> artifact, whitespace, bangs
- transducers
- :caching? :snapshot -> :cache
- assert :content in artifact
- dosync

matgis-king reviewed (changes requested):
Reviewed. Some concerns with the changes.

erikangelin-king committed:
DEFEDIT-1072 Review fixes fixes

- actually compare graphs for determining wheter to use system cache
- compare graphs by identical? instead of =
